--- src/Makefile.in.orig	2012-04-10 00:49:32.075309721 -0400
+++ src/Makefile.in	2012-04-10 00:49:58.417980390 -0400
@@ -87,6 +87,7 @@
 
 install: cleanbin
 	if [ ! -d $(CONFDIR) ]; then mkdir -p $(CONFDIR); fi
+	if [ ! -d $(DESTDIR)/@bindir@ ]; then mkdir -p $(DESTDIR)/@bindir@; fi
 	if [ ! -e $(CONFDIR)/reaver.db ]; then cp reaver.db $(CONFDIR)/reaver.db && chmod -R a+rw $(CONFDIR); fi
 	if [ -e wash ]; then cp wash $(DESTDIR)/@bindir@/wash; fi
 	if [ -e reaver ]; then cp reaver $(DESTDIR)/@bindir@/reaver; fi
