--- exchange.c~	2014-09-11 23:05:45.173398690 -0400
+++ exchange.c	2014-09-11 23:05:45.229399608 -0400
@@ -132,6 +132,11 @@
                 break;
             case M7:
                 cprintf(VERBOSE, "[+] Received M7 message\n");
+		if(!m6_sent)
+		{
+			tx_type = SEND_WSC_NACK;
+			terminated = 1;
+		}
                 /* Fall through */
             case DONE:
                 if(get_key_status() == KEY2_WIP) 
