--- exchange.c~	2014-09-12 00:40:09.532393118 -0400
+++ exchange.c	2014-09-12 00:40:09.576393817 -0400
@@ -240,7 +240,7 @@
          * Only treat the timeout as a NACK if this feature is enabled.
          */
         if(get_timeout_is_nack() &&
-                (last_msg == M3 || last_msg == M5))
+                ((last_msg == M3 && (get_key_status() == KEY1_WIP)) || last_msg == M5))
         {
             ret_val = KEY_REJECTED;
         }
