--- wpsmon.c~	2014-09-11 23:30:42.985770873 -0400
+++ wpsmon.c	2014-09-11 23:30:43.029771585 -0400
@@ -50,7 +50,7 @@
         { "probes", required_argument, NULL, 'n' },
         { "daemonize", no_argument, NULL, 'D' },
         { "file", no_argument, NULL, 'f' },
-        { "ignore-fcs", no_argument, NULL, 'C' },
+        { "announce-fcs", no_argument, NULL, 'C' },
         { "5ghz", no_argument, NULL, '5' },
         { "scan", no_argument, NULL, 's' },
         { "survey", no_argument, NULL, 'u' },
@@ -67,7 +67,7 @@
     set_auto_channel_select(0);
     set_wifi_band(BG_BAND);
     set_debug(INFO);
-    set_validate_fcs(1);
+    set_validate_fcs(0);
     set_log_file(stdout);
     set_max_num_probes(DEFAULT_MAX_NUM_PROBES);
 
@@ -104,7 +104,7 @@
                 mode = SURVEY;
                 break;
             case 'C':
-                set_validate_fcs(0);
+                set_validate_fcs(1);
                 break;
             case 'D':
                 daemonize();
@@ -427,7 +427,7 @@
     fprintf(stderr, "\t-o, --out-file=<file>                Write data to file\n");
     fprintf(stderr, "\t-n, --probes=<num>                   Maximum number of probes to send to each AP in scan mode [%d]\n", DEFAULT_MAX_NUM_PROBES);
     fprintf(stderr, "\t-D, --daemonize                      Daemonize wash\n");
-    fprintf(stderr, "\t-C, --ignore-fcs                     Ignore frame checksum errors\n");
+    fprintf(stderr, "\t-C, --announce-fcs                   Announce frame checksum errors\n");
     fprintf(stderr, "\t-5, --5ghz                           Use 5GHz 802.11 channels\n");
     fprintf(stderr, "\t-s, --scan                           Use scan mode\n");
     fprintf(stderr, "\t-u, --survey                         Use survey mode [default]\n");
