diff -urN compat-drivers-3.9-rc2-2.old/drivers/net/wireless/rt2x00/rt2x00mac.c compat-drivers-3.9-rc2-2/drivers/net/wireless/rt2x00/rt2x00mac.c
--- compat-drivers-3.9-rc2-2.old/drivers/net/wireless/rt2x00/rt2x00mac.c	2013-03-12 01:42:48.000000000 +0100
+++ compat-drivers-3.9-rc2-2/drivers/net/wireless/rt2x00/rt2x00mac.c	2013-03-14 15:44:10.522250488 +0100
@@ -141,7 +141,14 @@
 	 * either RTS or CTS-to-self frame and handles everything
 	 * inside the hardware.
 	 */
-	if (!rt2x00dev->ops->hw->set_rts_threshold &&
+	/*
+	 * We want to enable the manual use of RTS/CTS and CTS-to-self in
+	 * monitor mode.
+	 * Check for monitor mode here, but the vif seems to be null in monitor
+	 * mode.
+	 */
+	if ((!rt2x00dev->ops->hw->set_rts_threshold || !tx_info->control.vif ||
+	     tx_info->control.vif->type == NL80211_IFTYPE_MONITOR) &&
 	    (tx_info->control.rates[0].flags & (IEEE80211_TX_RC_USE_RTS_CTS |
 						IEEE80211_TX_RC_USE_CTS_PROTECT))) {
 		if (rt2x00queue_available(queue) <= 1)
