diff -rupN linux-source-3.16/drivers/net/wireless/ath/ath9k/common-init.c linux-new/drivers/net/wireless/ath/ath9k/common-init.c
--- linux-source-3.16/drivers/net/wireless/ath/ath9k/common-init.c	2015-03-30 12:09:09.000000000 +0200
+++ linux-new/drivers/net/wireless/ath/ath9k/common-init.c	2015-06-05 16:26:22.339916633 +0200
@@ -22,7 +22,7 @@
 	.band = IEEE80211_BAND_2GHZ, \
 	.center_freq = (_freq), \
 	.hw_value = (_idx), \
-	.max_power = 20, \
+	.max_power = 36, \
 }
 
 #define CHAN5G(_freq, _idx) { \
diff -rupN linux-source-3.16/drivers/net/wireless/ath/regd.c linux-new/drivers/net/wireless/ath/regd.c
--- linux-source-3.16/drivers/net/wireless/ath/regd.c	2015-03-30 12:09:09.000000000 +0200
+++ linux-new/drivers/net/wireless/ath/regd.c	2015-06-05 16:23:17.827924393 +0200
@@ -33,12 +33,12 @@ static int __ath_regd_init(struct ath_re
  */
 
 /* Only these channels all allow active scan on all world regulatory domains */
-#define ATH9K_2GHZ_CH01_11	REG_RULE(2412-10, 2462+10, 40, 0, 20, 0)
+#define ATH9K_2GHZ_CH01_11	REG_RULE(2412-10, 2462+10, 40, 0, 36, 0)
 
 /* We enable active scan on these a case by case basis by regulatory domain */
-#define ATH9K_2GHZ_CH12_13	REG_RULE(2467-10, 2472+10, 40, 0, 20,\
+#define ATH9K_2GHZ_CH12_13	REG_RULE(2467-10, 2472+10, 40, 0, 36,\
 					 NL80211_RRF_NO_IR)
-#define ATH9K_2GHZ_CH14		REG_RULE(2484-10, 2484+10, 40, 0, 20,\
+#define ATH9K_2GHZ_CH14		REG_RULE(2484-10, 2484+10, 40, 0, 36,\
 					 NL80211_RRF_NO_IR | \
 					 NL80211_RRF_NO_OFDM)
 
